### PR TITLE
REFAC(client): Replace <codecvt> usage

### PR DESCRIPTION
**Pull request by davidebeatrici**
Deprecated in C++17 and slated for removal in C++26.

Unfortunately no standard replacement is available yet.

### CODE REVIEW

File: plugins/se/se.cpp [hunk 1]
```diff
@@ -195,10 +195,12 @@ static bool tryInit(const std::multimap< std::wstring, unsigned long long int >
 	};
 
 	for (const auto &name : names) {
-		const auto id = ProcessBase::find(name, pids);
-		if (!id) {
-			continue;
+		const auto iter = pids.find(std::wstring(name.cbegin(), name.cend()));
+		if (iter == pids.cend()) {
+			return false;
```
Comment (Hartmnt) [inline plugins/se/se.cpp:200]: It seems to me that the code logic was modified here. Wouldn't we want to ``continue`` instead of ``return false;``?

File: src/mumble/GlobalShortcut_win.h [hunk 2]
```diff
@@ -134,6 +134,8 @@ public slots:
 	std::unique_ptr< GKeyLibrary > m_gkey;
 #endif
 	DeviceMap::iterator addDevice(const HANDLE deviceHandle);
+
+	static std::string utf16To8(const std::wstring &wstr);
```
Comment (Hartmnt) [inline src/mumble/GlobalShortcut_win.h:138]: Would it make sense to move this to a utility class in case we want to use it later in other parts of the code base?

File: plugins/se/se.cpp (hunk 1 above)
Comment (davidebeatrici) [reply plugins/se/se.cpp:200]: Yup, I accidentally copied the code from `cod2.cpp`.

File: src/mumble/GlobalShortcut_win.h (hunk 2 above)
Comment (davidebeatrici) [reply src/mumble/GlobalShortcut_win.h:138]: It's Windows-only. The only other instance of `WideCharToMultiByte()` is:

https://github.com/mumble-voip/mumble/blob/b75fe546ab6e5f32204e20df7ee3ceb8ce70262c/src/ProcessResolver.cpp#L67-L76

Which, now that I'm reading it, should be refactored to avoid the use of `MAX_PATH`.

I think I'll do that in a separate pull request (while also creating a class to avoid code duplication).

Comment (Krzmbrzl) [reply src/mumble/GlobalShortcut_win.h:138]: I don't see a reason for not doing that right away.

Comment (davidebeatrici) [reply src/mumble/GlobalShortcut_win.h]: `utf16To8()` now lives in `win.h`, `win` namespace.

I will create a separate pull request to refactor `ProcessResolver.cpp`.

**Review by Hartmnt — APPROVED** [review]

File: 3rdparty/utfcpp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+add_library(utfcpp INTERFACE)
```
Comment (Krzmbrzl) [inline 3rdparty/utfcpp/CMakeLists.txt:1]: Why don't we include this as a submodule?

File: plugins/HostWindows.cpp [hunk 4]
```diff
@@ -42,7 +40,7 @@ Modules HostWindows::modules() const {
 	MODULEENTRY32 me;
 	me.dwSize = sizeof(me);
 	for (auto ok = Module32First(snapshotHandle, &me); ok; ok = Module32Next(snapshotHandle, &me)) {
-		const auto name = utf16ToUtf8(reinterpret_cast< char16_t * >(me.szModule));
+		const auto name = utf16To8(me.szModule);
```
Comment (Krzmbrzl) [inline plugins/HostWindows.cpp:43]: Now that the cast is gone, we shouldn't use `auto` anymore

File: src/mumble/LegacyPlugin.cpp [hunk 5]
```diff
@@ -195,11 +194,26 @@ uint8_t LegacyPlugin::initPositionalData(const char *const *programNames, const
...
 		for (size_t i = 0; i < programCount; i++) {
-			std::string currentName = programNames[i];
-			std::wstring currentNameWstr =
-				std::wstring_convert< std::codecvt_utf8< wchar_t > >().from_bytes(currentName);
-
-			pidMap.insert(std::pair< std::wstring, unsigned long long int >(currentNameWstr, programPIDs[i]));
+			const std::string currentName = programNames[i];
+
+			try {
+				switch (sizeof(wchar_t)) {
```
Comment (Krzmbrzl) [inline src/mumble/LegacyPlugin.cpp:200]: I have seen this implementation in at least two places. Why not turn it into a function and reuse it?

File: src/win.h [hunk 6]
```diff
@@ -51,4 +53,26 @@ typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
 #	endif
 #endif
 
+namespace win {
+static inline std::string utf16To8(const std::wstring &wstr) {
```
Comment (Krzmbrzl) [inline src/win.h:57]: Why do we need a Windows-specific conversion function? Shouldn't the new library be able to do this just as well?

File: plugins/HostWindows.cpp (hunk 4 above)
Comment (davidebeatrici) [reply plugins/HostWindows.cpp:43]: Please note that the type was already "hidden" before. The cast was for the argument to `utf16ToUtf8()`.

File: 3rdparty/utfcpp/CMakeLists.txt (hunk 3 above)
Comment (davidebeatrici) [reply 3rdparty/utfcpp/CMakeLists.txt:1]: It's not the recommended way by upstream. Also, they are discontinuing the CMake project.

File: src/mumble/LegacyPlugin.cpp (hunk 5 above)
Comment (davidebeatrici) [reply src/mumble/LegacyPlugin.cpp:200]: I confirm there are two instances of it. Consider, though, that the stuff in `plugins` should be standalone (i.e. not depending directly on files from `src`).

File: 3rdparty/utfcpp/CMakeLists.txt (hunk 3 above)
Comment (Krzmbrzl) [reply 3rdparty/utfcpp/CMakeLists.txt:1]: In that case I would suggest forking the library under our organization and then using that as a submodule.

File: src/mumble/LegacyPlugin.cpp (hunk 5 above)
Comment (Krzmbrzl) [reply src/mumble/LegacyPlugin.cpp:200]: Alright - we should probably still create a dedicated function in both directories that can be reused by other code without having to rip it out of existing code.

File: plugins/HostWindows.cpp (hunk 4 above)
Comment (Krzmbrzl) [reply plugins/HostWindows.cpp]: True - then it is high time to remedy this issue :)

File: src/win.h (hunk 6 above)
Comment (Krzmbrzl) [reply src/win.h]: Then why the Windows-specific function? 👀

File: plugins/HostWindows.cpp (hunk 4 above)
Comment (davidebeatrici) [reply plugins/HostWindows.cpp]: Yeah, I did that.

File: src/mumble/LegacyPlugin.cpp (hunk 5 above)
Comment (davidebeatrici) [reply src/mumble/LegacyPlugin.cpp:200]: Keep in mind that Link is the only plugin that will ever need such a logic, for one precise reason: its API makes use of `wchar_t`, which is 2 bytes on Windows and 4 bytes on everything else.

Other plugins have to convert from the correct encoding based on the game they target.

Of course there may be cases where a game internally uses `wchar_t` and is built for both Linux and Windows. However, in such cases we cannot rely on `sizeof(wchar_t)` anyway because it would refer to the current process (Mumble client).

File: src/win.h (hunk 6 above)
Comment (davidebeatrici) [reply src/win.h]: I realized later that we needed the third-party (header-only) library in the client anyway, for `LegacyPlugin.cpp`.

As for `HostWindows.cpp`, I kept the Windows-specific function otherwise all plugins would've depended on the external library.

File: 3rdparty/utfcpp/CMakeLists.txt (hunk 3 above)
Comment (davidebeatrici) [reply 3rdparty/utfcpp/CMakeLists.txt]: Done.

File: src/win.h (hunk 6 above)
Comment (Krzmbrzl) [reply src/win.h]: If the library is only used in the implementation of that class, downstream code doesn't depend on it (it's a private dependency in that case)

Comment (davidebeatrici) [reply src/win.h]: Only if the class is compiled into a separate object. In our case it's compiled as part of the plugin.

Comment (Krzmbrzl) [reply src/win.h]: > Only if the class is compiled into a separate object.

which is the case as every translation unit ends up as an individual object file, no?

Comment (davidebeatrici) [reply src/win.h]: Yes, what I meant is that the object is not part of a library that the plugin links to.

In short: the third-party library has to exist in order for EVERY plugin (that relies on `HostWindows.cpp`) to be compiled. It's something I've always wanted to avoid, in fact the various classes don't depend on anything other than the standard library and the OS' API.

Comment (Krzmbrzl) [reply src/win.h]: I really don't see the issue here. We embed the source code of that library in Mumble anyway. And is people are using `HostWindows`, they depend on our source code anyway :shrug:

Comment (Krzmbrzl) [reply src/win.h]: Don't see a reason why that would be required/useful. Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself.
If we want separately buildable plugins, we should create a separate repo and put them there. :shrug:

Comment (davidebeatrici) [reply src/win.h]: To be clear: I wouldn't be opposed to rely on a library if it turned out to be necessary, but in this case we don't gain any relevant advantages.

Also, keep in mind that the functions offered by Windows are designed to work with the argument(s) that are commonly used throughout WinAPI.

`utfcpp` deliberately avoids `std::wstring` because it doesn't use the same encoding on all platforms. This is due to `wchar_t` being 2 bytes on Windows (UTF-16) and 4 bytes on UNIX (UTF-32).

Basically, we don't end up with cleaner code as casting is required.

Comment (davidebeatrici) [reply src/win.h]: > Don't see a reason why that would be required/useful. Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself. If we want separately buildable plugins, we should create a separate repo and put them there. 🤷

> Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself.

I wouldn't go as far as saying "exact same", but ABI compatibility is indeed required. It's one of the reasons we chose to write the new API in C rather than C++.

> If we want separately buildable plugins, we should create a separate repo and put them there. 🤷

That would be ideal.

Comment (Krzmbrzl) [reply src/win.h]: My point is that I would like to minimize the platform specific code in our code base as much as possible as that increases maintenance overhead.

Comment (davidebeatrici) [reply src/win.h]: Only if said code is required on multiple platforms. In this case it's a Windows-only conversion that is necessary.

File: src/mumble/GlobalShortcut_win.cpp [hunk 7]
```diff
@@ -22,7 +22,8 @@
 #	include "GKey.h"
 #endif
 
-#include <codecvt>
+#include "win.h"
```
Comment (davidebeatrici) [inline src/mumble/GlobalShortcut_win.cpp:25]: Left by mistake.
